fix: improve drag-and-drop behavior for tab folders (#11326)

garrza opened the pull request:
- Simplify drop zone logic with 25% edge zones for better UX
- Tabs can now be inserted into folder from either above or below, more forgiving drop as well
- Remove unnecessary upper/lower threshold preferences


https://github.com/user-attachments/assets/b935b485-d3d7-4ca6-ac5c-213500f03d43


Fixes #11326

garrza committed:
fix: improve drag-and-drop behavior for tab folders (#11326)

- Simplify drop zone logic with 25% edge zones for better UX
- Remove unnecessary upper/lower threshold preferences

Fixes #11326

mr-cheffy commented on src/zen/folders/ZenFolders.mjs:
```diff
@@ -899,7 +899,11 @@
+        if (
+          !group.collapsed ||
+          !Services.prefs.getBoolPref('zen.folders.search.enabled') ||
+          document.getElementById('tabbrowser-tabs')?.hasAttribute('movingtab')
```
Use `gBrowser.tabContainer` and it'll never be undefined

garrza replied:
nice, will do this.

mr-cheffy commented on src/zen/folders/ZenFolders.mjs:
```diff
@@ -1164,24 +1164,17 @@
-        !isSplitGroup;
-      const shouldDropNear = overlapPercent < dragUpThreshold || overlapPercent > dragDownThreshold;
+      // Edge zones: 25% top/bottom for dropping before/after, 50% middle for dropping inside
+      const edgeZone = 0.25;
```
I'd make this a preference

garrza replied:
yeah this sounds good, will make it so.

mr-cheffy commented on src/zen/folders/zen-folders.css:
```diff
@@ -116,6 +130,10 @@ zen-folder {
         }
 
+        #tabbrowser-tabs[movingtab] & {
+          transition: transform 0.15s ease, opacity 0.15s ease !important;
```
Why did you add this?

mr-cheffy commented on src/zen/folders/zen-folders.css:
```diff
@@ -45,6 +45,20 @@ zen-folder {
   }
 
+  #tabbrowser-tabs[movingtab] & {
+    contain: layout style;
```
Why's this necessary?

garrza replied:
not necessary, there is actually no need to change css, i'll revert these.

mr-cheffy commented:
https://github.com/user-attachments/assets/b4941751-77d6-4fe4-83d6-9c1ff738d0e0

now the folder doesn't transition though?

garrza commented:
simplified a bit by removing non-needed css, also added edge zone threshold as a preference loaded at component level.

added defaults for both max subfolders and edge zone threshold to handle timing issues during module initialization. module-level constants load before the preference system is fully ready, so defaults prevent `NS_ERROR_UNEXPECTED` errors that would break the entire `ZenFolders` module. 

no default for max subfolders preference could have potentially randomly broken the module for new users which do not have preferences loaded due to race condition.

mr-cheffy commented on src/zen/folders/ZenFolders.mjs:
```diff
@@ -29,7 +29,9 @@
-  const ZEN_MAX_SUBFOLDERS = Services.prefs.getIntPref('zen.folders.max-subfolders');
+  const ZEN_MAX_SUBFOLDERS = Services.prefs.getIntPref('zen.folders.max-subfolders', 5);
+  const ZEN_EDGE_ZONE_THRESHOLD =
+    Services.prefs.getIntPref('zen.view.drag-and-drop.edge-zone-threshold', 25) / 100;
```
move these constants inside the nsZenFolders, e.g. `#ZEN_MAX_SUBFOLDERS`, ...

garrza replied:
gotcha, lemme do that man

mr-cheffy committed:
Merge branch 'dev' into fix/tab-folder-drag-drop

mr-cheffy commented:
Please, don't push with `--force` as its more difficult for me to update this branch

mr-cheffy reviewed (approved):
Looks good! Lets drop this on twilight and i'll start testing on how it generally feels and different configuration options!
